Handle duplicates in test names

PerMac opened the pull request:
Several duplicates of test names where found in zephyr tree. This is problematic since tests with the same name will be overwritten in the result report 
This PR consist of 2 commits:
- resolving found duplicates
- by default, twister will raise an error whenever duplicates are found

PerMac committed:
tests: samples: Resolve duplicates in names.

Several duplicates were found with
scripts/twister -T samples/ -T tests/ --dry-run --list-test-duplicates
This is an issue since duplicated names causes overwriting of
results. Most duplicates looked like obvious copy-pase errors.
New names where addopted looking at other tests in the same yaml
or looking at the directory/descriptio.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

Thalley commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -113,6 +113,10 @@ def add_parse_arguments(parser = None):
                              help="""List tests with duplicate identifiers.
         """)
 
+    case_select.add_argument("--test-duplicates-as-error", action="store_true",
+                             help="""Terminate with error when duplicate identifiers.
+        """)
```
```suggestion
    case_select.add_argument("--test-duplicates-as-error", action="store_true",
                             help="""Terminate with error when duplicate identifiers.""")
```

stephanosio commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -113,6 +113,10 @@ def add_parse_arguments(parser = None):
                              help="""List tests with duplicate identifiers.
         """)
 
+    case_select.add_argument("--test-duplicates-as-error", action="store_true",
+                             help="""Terminate with error when duplicate identifiers.
+        """)
```
Should this default to true? IMHO, twister should have always been treating this as an error.

PerMac replied:
I also think it should be default. However, I don't want to break CIs of other people, since those can contain duplicates in tests not in zephyr. The last commit adds this flag to zephyr's on-pr checks, to prevent new duplicates being added to tree.
The option --list-test-duplicates is available for a while. Maybe there was another reason why duplicates are not errors by default? @nashif?

nashif replied:
let's make it the default.

hakehuang commented:
@PerMac , just curious on how do you found the duplicated cases name? is there a script? if so, maybe we can add twist a build in test?

PerMac commented:
> @PerMac , just curious on how do you found the duplicated cases name? is there a script? if so, maybe we can add twist a build in test?

call twister with `--list-test-duplicates`. Last commit adds this flag to on-pr CIs call.

nashif commented on scripts/pylib/twister/twisterlib/environment.py:
```diff
@@ -113,6 +113,9 @@ def add_parse_arguments(parser = None):
         """)
 
+    case_select.add_argument("--test-duplicates-as-error", action="store_true",
+                             help="""Terminate with error when duplicate identifiers.""")
```
This should be default behavior and we should not require yet another option, let's just fail by default.

PerMac replied:
@nashif Not clear.  Should I get rid of the option completely and have it always active in the code? If so, is the change from comment  `replace with "Treat duplicate identifiers as errors."` no longer needed?

nashif replied:
> Should I get rid of the option completely and have it always active in the code?

yes

>  If so, is the change from comment replace with "Treat duplicate identifiers as errors." no longer needed?

yes, not needed.

nashif commented on scripts/pylib/twister/twisterlib/testplan.py:
```diff
@@ -289,9 +289,16 @@ def handle_modules(self):
-        if self.options.list_test_duplicates:
-            self.report_duplicates()
-            return 0
+        if self.options.list_test_duplicates or self.options.test_duplicates_as_error:
```
instead of using the report workflow, it would make sense to do that during discovery, i.e.:

```
diff --git a/scripts/pylib/twister/twisterlib/testplan.py b/scripts/pylib/twister/twisterlib/testplan.py
index ad2486e4b0..818ca30b25 100755
--- a/scripts/pylib/twister/twisterlib/testplan.py
+++ b/scripts/pylib/twister/twisterlib/testplan.py
@@ -175,6 +175,11 @@ class TestPlan:
         for _, ts in self.testsuites.items():
             self.scenarios.append(ts.id)

+        seen = set()
+        dupes = [x for x in self.scenarios if x in seen or seen.add(x)]
+        if dupes:
+            raise TwisterRuntimeError(f"Duplicate test scenarios found: {dupes}.")
+
         self.parse_configuration(config_file=self.env.test_config)
         self.add_configurations()

```

PerMac committed:
twister: Raise error when duplicates found on default

Make twister to raise error by default whenever duplicated test ids
are found. Remove redundant option to list duplicates.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented:
@nashif Done as you suggested:
- Twister will raise an error by default whenever duplicates are discovered during discovery phase.
- For this, I reused the existing (slightly modified logging) method report_duplicates()
- I removed the flag `--list-test-duplicates` since now duplicates are listed by default

PerMac commented:
New batch of duplicates detected and caused the failure, updated with a fix
```
sample.bluetooth.periodic_adv found in:
  - /__w/zephyr/zephyr/samples/bluetooth/periodic_adv
  - /__w/zephyr/zephyr/samples/bluetooth/periodic_adv_conn
  - /__w/zephyr/zephyr/samples/bluetooth/periodic_adv_rsp
- sample.bluetooth.periodic_sync found in:
  - /__w/zephyr/zephyr/samples/bluetooth/periodic_sync
  - /__w/zephyr/zephyr/samples/bluetooth/periodic_sync_conn
  - /__w/zephyr/zephyr/samples/bluetooth/periodic_sync_rsp
- sample.usbc.sink found in:
  - /__w/zephyr/zephyr/samples/subsys/usb_c/sink
  - /__w/zephyr/zephyr/samples/subsys/usb_c/source
  ```
